File: examples/python/momentum_bot.py
#!/usr/bin/env python3
"""CoinRithm momentum bot (Python) -- a complete, runnable futures-agent template.

The Python twin of examples/bots/momentum-bot.mjs, built on the zero-dependency
client in coinrithm.py (standard library only).

What it does:
  1. whoami                  -- confirm the key + scopes
  2. resolve(SYMBOL)         -- symbol -> coinId (UCID)
  3. market(coinId)          -- 1h/24h momentum signal (long/short/none)
  4. futures_quote(...)      -- entry price, liquidation price, eligibility
  5. Confirm gate            -- DEFAULT IS A DRY RUN: prints the exact plan and
                                exits. Set LIVE=1 to actually open.
  6. open_futures(...)       -- with stop-loss + take-profit set at open
  7. Poll trades(updated_since=cursor) -- persisted cursor in a .state.json
     next to this file; dedupe by (venue, id); back off on 429 (the client
     honors Retry-After)
  8. Report when the stop-loss / take-profit / liquidation fires.

How to run:
  COINRITHM_API_KEY=crk_live_xxx python3 examples/python/momentum_bot.py

Env (all optional except the key):
  LIVE=1         actually place the trade. Default: DRY RUN -- trades NOTHING.
  SYMBOL         default BTC
  LEVERAGE       default 2 (max 20)
  MARGIN_MUSD    default 50 (min 10)
  SL_PCT/TP_PCT  stop/target distance from entry, default 2 / 4 (%)
  POLL_SECONDS   default 60       MAX_POLLS  default 120 (then exits; the
                 cursor is persisted, so re-running RESUMES the watch)

What it costs: nothing. PAPER trading -- a 50,000 virtual-mUSD account.
No real money, no real exchange, not financial advice.
"""

import json
import os
import pathlib
import sys
import time

sys.path.insert(0, str(pathlib.Path(__file__).parent))
from coinrithm import CoinRithm  # noqa: E402

LIVE = os.environ.get("LIVE") == "1"
SYMBOL = os.environ.get("SYMBOL", "BTC")
LEVERAGE = min(float(os.environ.get("LEVERAGE", 2)), 20)
MARGIN = max(float(os.environ.get("MARGIN_MUSD", 50)), 10)
SL_PCT = float(os.environ.get("SL_PCT", 2)) / 100
TP_PCT = float(os.environ.get("TP_PCT", 4)) / 100
POLL_SECONDS = int(os.environ.get("POLL_SECONDS", 60))
MAX_POLLS = int(os.environ.get("MAX_POLLS", 120))

STATE_FILE = pathlib.Path(__file__).parent / ".momentum_bot.state.json"


def load_state() -> dict:
    try:
        return json.loads(STATE_FILE.read_text())
    except (OSError, ValueError):
        return {"cursor": None, "seen": [], "position": None}


def save_state(state: dict) -> None:
    STATE_FILE.write_text(json.dumps(state, indent=2))


def watch_position(crk: CoinRithm, state: dict) -> None:
    """Delta-poll /trades with the persisted asOf cursor until our position
    closes (stop_loss / take_profit / liquidation / user_close)."""
    pos = state["position"]
    print(f"\nWatching position {pos['id']} ({pos['side']} {pos['symbol']}) -- "
          f"SL={pos.get('stopLossPrice')} TP={pos.get('takeProfitPrice')}")
    for i in range(MAX_POLLS):
        if state["cursor"]:
            j = crk.trades(venue="futures", updated_since=state["cursor"])
        else:
            j = crk.trades(venue="futures", limit=1)
        state["cursor"] = j["asOf"]  # ALWAYS advance the cursor from the response
        fresh = [t for t in j.get("trades", []) if f"{t['venue']}:{t['id']}" not in state["seen"]]
        state["seen"] = (state["seen"] + [f"{t['venue']}:{t['id']}" for t in fresh])[-200:]
        save_state(state)

        if fresh:
            mine = next((p for p in crk.futures_positions().get("positions", []) if p["id"] == pos["id"]), None)
            if mine and mine["status"] != "open":
                pnl = mine.get("realizedPnlMusd") or 0
                print(f"\nPosition {pos['id']} closed: exitReason={mine.get('exitReason')} "
                      f"exit={mine.get('exitPrice')} realizedPnl={pnl:+.2f} mUSD")
                state["position"] = None
                save_state(state)
                return
            print(f"  poll {i + 1}: {len(fresh)} new closed trade(s), ours still open")
        else:
            print(f"  poll {i + 1}: no new fills (cursor={state['cursor']})")
        time.sleep(POLL_SECONDS)
    print(f"\nReached MAX_POLLS={MAX_POLLS}. Cursor + position persisted -- re-run to resume.")


def main() -> None:
    key = os.environ.get("COINRITHM_API_KEY") or os.environ.get("CRK_API_KEY")
    if not key:
        sys.exit("Set COINRITHM_API_KEY (CoinRithm -> Profile -> API Keys).")
    crk = CoinRithm(api_key=key, base_url=os.environ.get("BASE_URL", "https://api.coinrithm.com"))

    mode = "LIVE (paper trades WILL be placed)" if LIVE else "DRY RUN (plan only -- set LIVE=1 to trade)"
    print(f"\n=== CoinRithm momentum bot (Python) -- {mode} ===")
    print("Paper trading only: virtual mUSD, no real money.\n")

    state = load_state()
    if state.get("position"):
        print(f"Resuming: state file has open position {state['position']['id']}.")
        watch_position(crk, state)
        return

    # 1) Identity + scopes
    me = crk.whoami()
    scopes = set(me.get("scopes", []))
    print(f"1) whoami: user {me['userId']}, scopes {sorted(scopes)}")
    if LIVE and "trade:futures" not in scopes:
        sys.exit("LIVE=1 needs a key with trade:futures scope.")

    # 2) Resolve the symbol to a coinId (UCID)
    match = crk.resolve(SYMBOL).get("match")
    if not match:
        sys.exit(f'could not resolve "{SYMBOL}" to a coin')
    coin_id = match["coinId"]
    print(f"2) resolve: {SYMBOL} -> coinId {coin_id} ({match['name']})")

    # 3) Momentum signal (a TEACHING heuristic, not alpha)
    m = crk.market(coin_id)
    price = m.get("price") or {}
    c1, c24 = price.get("change1h") or 0, price.get("change24h") or 0
    print(f"3) market: ${price.get('usd')} | 1h {c1}% 24h {c24}%")
    side = "long" if (c1 > 0 and c24 > 1) else "short" if (c1 < 0 and c24 < -1) else None
    if not side:
        print("   No signal (1h and 24h momentum must agree, |24h| > 1%). Exiting.")
        return
    print(f"   Signal: {side.upper()}")

    # 4) Quote first (read-only)
    q = crk.futures_quote(coin_id, side, leverage=LEVERAGE, margin_musd=MARGIN)
    if not q.get("eligible"):
        print(f"4) quote: entry BLOCKED -- {q.get('blockReasons')}. Exiting.")
        return
    entry, liq = q["entryPrice"], q["liquidationPrice"]
    print(f"4) quote: entry~{entry} liq~{liq} notional={q.get('notionalMusd')} mUSD")

    # 5) Side-aware SL/TP corridor (long: liq < SL < entry < TP; short inverted)
    if side == "long":
        sl = max(entry * (1 - SL_PCT), liq + (entry - liq) * 0.1)
        tp = entry * (1 + TP_PCT)
    else:
        sl = min(entry * (1 + SL_PCT), liq - (liq - entry) * 0.1)
        tp = entry * (1 - TP_PCT)
    sl, tp = round(sl, 6), round(tp, 6)

    plan = {"coinId": coin_id, "symbol": match["symbol"], "side": side, "leverage": LEVERAGE,
            "marginMusd": MARGIN, "entryPrice": entry, "stopLossPrice": sl,
            "takeProfitPrice": tp, "liquidationPrice": liq}
    print(f"5) plan:\n{json.dumps(plan, indent=2)}")

    if not LIVE:
        t = crk.trades(venue="futures", limit=1)
        state["cursor"] = t["asOf"]
        save_state(state)
        print(f"\nDRY RUN -- no order placed. (Captured asOf cursor {t['asOf']} into {STATE_FILE.name}.)")
        print("Set LIVE=1 to open this position with the SL/TP above (paper funds only).")
        return

    # 6) Open WITH SL/TP set atomically (idempotency key: unique per intent)
    res = crk.open_futures(coin_id, side, leverage=LEVERAGE, margin_musd=MARGIN,
                           idempotency_key=CoinRithm.new_idempotency_key("momentum"),
                           stop_loss_price=sl, take_profit_price=tp)
    pos = res["position"]
    print(f"6) opened position {pos['id']}: {pos['side']} {pos['sizeCoin']} {match['symbol']} "
          f"@ {pos['entryPrice']} (SL {pos['stopLossPrice']}, TP {pos['takeProfitPrice']})")

    # 7) Capture a fresh cursor, persist, and watch for the SL/TP to fire
    t = crk.trades(venue="futures", limit=1)
    state["cursor"] = t["asOf"]
    state["position"] = {"id": pos["id"], "side": side, "symbol": match["symbol"],
                         "stopLossPrice": pos.get("stopLossPrice"), "takeProfitPrice": pos.get("takeProfitPrice")}
    save_state(state)
    watch_position(crk, state)


if __name__ == "__main__":
    main()

File: examples/bots/momentum-bot.mjs
#!/usr/bin/env node
// ---------------------------------------------------------------------------
// CoinRithm MOMENTUM BOT — a complete, runnable futures-agent template.
//
// What it does (one full agent loop, ~150 lines of logic):
//   1. whoami                  — confirm the key + scopes
//   2. GET /resolve?q=SYMBOL   — symbol -> coinId (UCID)
//   3. GET /market/:coinId     — 1h/24h/7d momentum signal (long/short/none)
//   4. POST /futures/quote     — entry price, liquidation price, eligibility
//   5. Confirm gate            — DEFAULT IS A DRY RUN: prints the exact plan
//                                and exits. Set LIVE=1 to actually open.
//   6. POST /futures/open      — with stopLossPrice + takeProfitPrice set
//                                atomically at open (side-aware corridor)
//   7. Poll GET /trades?updatedSince=<asOf> — persisted cursor in a
//      .state.json next to this file; dedupe by (venue,id); paced by
//      RateLimit-Remaining; backs off on 429 Retry-After
//   8. Report when the stop-loss / take-profit / liquidation fires
//      (exitReason + realized PnL), then check the public Agent Arena.
//
// How to run:
//   COINRITHM_API_KEY=crk_live_xxx node examples/bots/momentum-bot.mjs
//
// Env:
//   COINRITHM_API_KEY  (required)  key scopes: read (+ trade:futures for LIVE)
//   LIVE=1             (optional)  actually place the trade. Default: DRY RUN —
//                                  reads + quote only, prints the plan, trades NOTHING.
//   SYMBOL             (optional)  default BTC
//   LEVERAGE           (optional)  default 2 (max 20)
//   MARGIN_MUSD        (optional)  default 50 (min 10)
//   SL_PCT / TP_PCT    (optional)  stop/target distance from entry, default 2 / 4 (%)
//   POLL_SECONDS       (optional)  default 60 (the SL/TP worker runs per-minute)
//   MAX_POLLS          (optional)  default 120 (~2h), then exits; the cursor is
//                                  persisted, so re-running RESUMES the watch.
//   BASE_URL           (optional)  default https://api.coinrithm.com
//
// What it costs: nothing. This is PAPER trading — a 50,000 virtual-mUSD
// account. No real money, no real exchange, not financial advice.
//
// Zero dependencies — Node 18+ (built-in fetch) only.
// ---------------------------------------------------------------------------

import { readFileSync, writeFileSync } from "node:fs";
import { randomUUID } from "node:crypto";
import { fileURLToPath } from "node:url";
import path from "node:path";

// One runId per bot session — groups all ledger rows for this run so the
// evidence bundle is exportable via GET /api/agent/ledger/export?runId=RUN_ID.
const RUN_ID = `momentum-${new Date().toISOString().slice(0, 10)}-${randomUUID().slice(0, 8)}`;

const BASE = process.env.BASE_URL || "https://api.coinrithm.com";
const KEY = process.env.COINRITHM_API_KEY || process.env.CRK_API_KEY;
const LIVE = process.env.LIVE === "1";
const SYMBOL = process.env.SYMBOL || "BTC";
const LEVERAGE = Math.min(Number(process.env.LEVERAGE) || 2, 20);
const MARGIN = Math.max(Number(process.env.MARGIN_MUSD) || 50, 10);
const SL_PCT = (Number(process.env.SL_PCT) || 2) / 100;
const TP_PCT = (Number(process.env.TP_PCT) || 4) / 100;
const POLL_SECONDS = Number(process.env.POLL_SECONDS) || 60;
const MAX_POLLS = Number(process.env.MAX_POLLS) || 120;

const STATE_FILE = path.join(
  path.dirname(fileURLToPath(import.meta.url)),
  ".momentum-bot.state.json",
);

if (!KEY) {
  console.error("Set COINRITHM_API_KEY (CoinRithm -> Profile -> API Keys).");
  process.exit(1);
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// --- agentTrace builder: one runId, per-call decisionId -------------------
const trace = (label, confidence) => ({
  runId: RUN_ID,
  decisionId: `${label}-${randomUUID().slice(0, 8)}`,
  strategyLabel: "momentum-futures",
  ...(confidence !== undefined ? { confidence } : {}),
});

// --- API helper: rate-limit pacing + 429 Retry-After backoff ---------------
const api = async (method, p, body) => {
  for (let attempt = 0; attempt < 3; attempt++) {
    const res = await fetch(`${BASE}${p}`, {
      method,
      headers: { Authorization: `Bearer ${KEY}`, "content-type": "application/json" },
      body: body ? JSON.stringify(body) : undefined,
    });
    if (res.status === 429) {
      const wait = Number(res.headers.get("retry-after")) || 30;
      console.log(`  … 429 rate-limited, backing off ${wait}s`);
      await sleep(wait * 1000);
      continue;
    }
    // Proactive pacing: if the per-key budget is nearly spent, wait for reset.
    const remaining = Number(res.headers.get("ratelimit-remaining"));
    if (Number.isFinite(remaining) && remaining < 8) {
      const reset = Number(res.headers.get("ratelimit-reset")) || 10;
      console.log(`  … RateLimit-Remaining=${remaining}, pacing ${reset}s`);
      await sleep(reset * 1000);
    }
    let json = null;
    try { json = await res.json(); } catch {}
    return { status: res.status, json };
  }
  throw new Error("rate-limited 3 times in a row — giving up");
};

const must = (r, what) => {
  if (r.status >= 400) throw new Error(`${what}: HTTP ${r.status} ${JSON.stringify(r.json)?.slice(0, 200)}`);
  return r.json;
};

// --- Persisted state: asOf cursor + (venue,id) dedupe + open position ------
const loadState = () => {
  try { return JSON.parse(readFileSync(STATE_FILE, "utf8")); } catch { return { cursor: null, seen: [], position: null }; }
};
const saveState = (s) => writeFileSync(STATE_FILE, JSON.stringify(s, null, 2));

// --- Settlement watch: delta-poll /trades with the asOf cursor -------------
// Stops, take-profits, and liquidations fire SERVER-SIDE (per-minute worker).
// The contract: pass the previous response's asOf back as updatedSince.
// Delivery is at-least-once, so dedupe by (venue,id).
const watchPosition = async (state) => {
  const pos = state.position;
  console.log(`\nWatching position ${pos.id} (${pos.side} ${pos.symbol}) — SL=${pos.stopLossPrice ?? "-"} TP=${pos.takeProfitPrice ?? "-"}`);
  console.log(`(polling /trades?updatedSince every ${POLL_SECONDS}s; cursor persisted in ${path.basename(STATE_FILE)})`);

  for (let i = 0; i < MAX_POLLS; i++) {
    const qs = state.cursor ? `?venue=futures&updatedSince=${encodeURIComponent(state.cursor)}` : "?venue=futures&limit=1";
    const j = must(await api("GET", `/api/agent/trades${qs}`), "GET /trades");
    state.cursor = j.asOf; // ALWAYS advance the cursor from the response
    const fresh = (j.trades || []).filter((t) => !state.seen.includes(`${t.venue}:${t.id}`));
    for (const t of fresh) state.seen.push(`${t.venue}:${t.id}`);
    state.seen = state.seen.slice(-200);
    saveState(state);

    if (fresh.length > 0) {
      // Something closed since the last poll — check whether it was OUR position.
      const pj = must(await api("GET", "/api/agent/positions/futures"), "GET /positions/futures");
      const mine = (pj.positions || []).find((p) => p.id === pos.id);
      if (mine && mine.status !== "open") {
        const pnl = mine.realizedPnlMusd ?? 0;
        console.log(`\n${pnl >= 0 ? "🟢" : "🔴"} Position ${pos.id} closed: exitReason=${mine.exitReason} exit=${mine.exitPrice} realizedPnl=${pnl.toFixed(2)} mUSD`);
        state.position = null;
        saveState(state);
        await arenaCheck();
        return;
      }
      console.log(`  poll ${i + 1}: ${fresh.length} new closed trade(s), ours still open`);
    } else {
      console.log(`  poll ${i + 1}: no new fills (cursor=${state.cursor})`);
    }
    await sleep(POLL_SECONDS * 1000);
  }
  console.log(`\nReached MAX_POLLS=${MAX_POLLS}. Cursor + position persisted — re-run to resume watching.`);
};

// --- Run-evidence export: prove the agent only acted on available data ------
// Response shape (with runId): { apiKeyId, exportedAt, count, maxRows,
//   run: { evidenceChecklist: { schema, overallStatus, items:[{id,label,status,detail}] },
//          executionAssumptions: { costModel, ... }, outcomeSummary, summary, ... },
//   data: [...ledger rows] }
const exportRunEvidence = async () => {
  try {
    const bundle = must(await api("GET", `/api/agent/ledger/export?runId=${encodeURIComponent(RUN_ID)}`), "GET /ledger/export");
    const run = bundle.run;
    const ec = run?.evidenceChecklist;
    console.log(`\nRun evidence (runId=${RUN_ID}):`);
    console.log(`  evidenceChecklist.overallStatus = ${ec?.overallStatus ?? "n/a"}`);
    for (const item of (ec?.items ?? [])) {
      console.log(`    [${item.status}] ${item.label}: ${item.detail}`);
    }
    console.log(`  rows in export    = ${bundle.count ?? bundle.data?.length ?? 0}`);
    const ea = run?.executionAssumptions;
    if (ea) console.log(`  costModel         = ${ea.costModel ?? "paper, mid/last price, no commission/slippage/funding (v1)"}`);
  } catch (e) {
    console.log(`  (run evidence unavailable: ${e.message})`);
  }
};

// --- Arena: how does this agent rank publicly? ------------------------------
const arenaCheck = async () => {
  const me = must(await api("GET", "/api/agent/me"), "GET /me");
  const perf = must(await api("GET", "/api/agent/performance"), "GET /performance");
  console.log(`\nScorecard: realized total ${perf.totals?.realizedPnlMusd?.toFixed?.(2) ?? "?"} mUSD over ${perf.totals?.tradeCount ?? 0} trades (winRate ${perf.totals?.winRate ?? "n/a"})`);
  if (!me.agentName) {
    console.log("Not on the Arena yet — set agentName + agentPublic on your key (Profile -> API Keys) to get ranked (any agent with at least one decided trade is ranked).");
    await exportRunEvidence();
    return;
  }
  const arena = must(await api("GET", "/api/arena?pageSize=50"), "GET /api/arena");
  const row = (arena.rows || []).find((r) => r.agentName === me.agentName);
  if (row) console.log(`Arena: #${row.rank} "${row.agentName}" — ${row.realizedPnlMusd.toFixed(2)} mUSD realized, winRate=${row.winRate ?? "n/a"}`);
  else console.log(`"${me.agentName}" not in the Arena top ${arena.rows?.length ?? 0} yet (needs ${arena.minDecidedTrades} decided trades + agentPublic).`);
  await exportRunEvidence();
};

// --- Main -------------------------------------------------------------------
const run = async () => {
  console.log(`\n=== CoinRithm momentum bot — ${LIVE ? "LIVE (paper trades WILL be placed)" : "DRY RUN (plan only — set LIVE=1 to trade)"} ===`);
  console.log(`Paper trading only: virtual mUSD, no real money.\n`);

  const state = loadState();

  // Resume: if a previous run left a position open, just keep watching it.
  if (state.position) {
    console.log(`Resuming: state file has open position ${state.position.id}.`);
    await watchPosition(state);
    return;
  }

  // 1) Identity + scopes
  const me = must(await api("GET", "/api/agent/me"), "GET /me");
  const scopes = new Set(me.scopes || []);
  console.log(`1) whoami: user ${me.userId}, scopes [${[...scopes].join(", ")}]`);
  if (LIVE && !scopes.has("trade:futures")) {
    console.error("LIVE=1 needs a key with trade:futures scope. Mint one in Profile -> API Keys.");
    process.exit(1);
  }

  // 2) Resolve the symbol to a coinId (UCID) — never pass tickers as coinId
  const res = must(await api("GET", `/api/agent/resolve?q=${encodeURIComponent(SYMBOL)}`), "GET /resolve");
  if (!res.match) throw new Error(`could not resolve "${SYMBOL}" to a coin`);
  const coinId = res.match.coinId;
  console.log(`2) resolve: ${SYMBOL} -> coinId ${coinId} (${res.match.name})`);

  // 3) Momentum signal from market context (a TEACHING heuristic, not alpha:
  //    trade only when 1h and 24h momentum agree and 24h move is > 1%)
  const m = must(await api("GET", `/api/agent/market/${coinId}`), "GET /market");
  const { change1h, change24h, change7d } = m.price || {};
  console.log(`3) market: $${m.price?.usd} | 1h ${change1h}% 24h ${change24h}% 7d ${change7d}%`);
  let side = null;
  if (change1h > 0 && change24h > 1) side = "long";
  else if (change1h < 0 && change24h < -1) side = "short";
  if (!side) {
    console.log("   No signal (1h and 24h momentum must agree, |24h| > 1%). Nothing to do — exiting.");
    return;
  }
  console.log(`   Signal: ${side.toUpperCase()}`);

  // 4) Quote first — read-only; shows entry, liquidation, eligibility.
  //    Check observation.freshness — stale data must not drive a trade.
  const q = must(await api("POST", "/api/agent/futures/quote", {
    coinId, side, leverage: LEVERAGE, marginMusd: MARGIN,
    agentTrace: trace("futures-quote", 0.6),
  }), "POST /futures/quote");
  const freshness = q.observation?.freshness;
  console.log(`4) quote: observation freshness=${freshness?.status ?? "n/a"} ageSeconds=${freshness?.ageSeconds ?? "n/a"} hash=${q.observation?.hash ?? "n/a"}`);
  if (freshness?.status && freshness.status !== "fresh") {
    console.log(`   Quote data is ${freshness.status} — skipping to avoid stale-data trade.`);
    return;
  }
  if (!q.eligible) {
    console.log(`4) quote: entry BLOCKED — ${JSON.stringify(q.blockReasons)}. Exiting.`);
    return;
  }
  const entry = q.entryPrice;
  const liq = q.liquidationPrice;
  console.log(`4) quote: entry≈${entry} liq≈${liq} notional=${q.notionalMusd} mUSD`);

  // 5) Side-aware SL/TP corridor (long: liq < SL < entry < TP; short inverted),
  //    clamped so the stop never lands on the wrong side of liquidation.
  let stopLossPrice, takeProfitPrice;
  if (side === "long") {
    stopLossPrice = Math.max(entry * (1 - SL_PCT), liq + (entry - liq) * 0.1);
    takeProfitPrice = entry * (1 + TP_PCT);
  } else {
    stopLossPrice = Math.min(entry * (1 + SL_PCT), liq - (liq - entry) * 0.1);
    takeProfitPrice = entry * (1 - TP_PCT);
  }
  stopLossPrice = Number(stopLossPrice.toPrecision(6));
  takeProfitPrice = Number(takeProfitPrice.toPrecision(6));

  const plan = { coinId, symbol: res.match.symbol, side, leverage: LEVERAGE, marginMusd: MARGIN, entryPrice: entry, stopLossPrice, takeProfitPrice, liquidationPrice: liq };
  console.log(`5) plan:\n${JSON.stringify(plan, null, 2)}`);

  if (!LIVE) {
    // DRY RUN ends here — but still demonstrate the sync cursor contract once.
    const t = must(await api("GET", "/api/agent/trades?venue=futures&limit=1"), "GET /trades");
    state.cursor = t.asOf;
    saveState(state);
    console.log(`\nDRY RUN — no order placed. (Captured /trades asOf cursor ${t.asOf} into ${path.basename(STATE_FILE)}.)`);
    console.log("Set LIVE=1 to open this position with the SL/TP above (paper funds only).");
    return;
  }

  // 6) Open WITH SL/TP set atomically. idempotencyKey is REQUIRED and unique
  //    per intent — a retry of the SAME intent must reuse the SAME key.
  const idempotencyKey = `momentum-${coinId}-${side}-${randomUUID()}`;
  const open = must(
    await api("POST", "/api/agent/futures/open", {
      coinId, side, leverage: LEVERAGE, marginMusd: MARGIN,
      idempotencyKey, stopLossPrice, takeProfitPrice,
      agentTrace: trace("futures-open", 0.6),
    }),
    "POST /futures/open",
  );
  const pos = open.position;
  console.log(`6) opened position ${pos.id}: ${pos.side} ${pos.sizeCoin} ${res.match.symbol} @ ${pos.entryPrice} (SL ${pos.stopLossPrice}, TP ${pos.takeProfitPrice})`);

  // 7) Capture a fresh cursor, persist, and watch for the SL/TP to fire.
  const t = must(await api("GET", "/api/agent/trades?venue=futures&limit=1"), "GET /trades");
  state.cursor = t.asOf;
  state.position = { id: pos.id, side, symbol: res.match.symbol, stopLossPrice: pos.stopLossPrice, takeProfitPrice: pos.takeProfitPrice };
  saveState(state);
  await watchPosition(state);
};

run().catch((e) => { console.error("FATAL", e.message); process.exit(1); });
